lib/flex.js

```diff
@@ -0,0 +1,132 @@
+const Configuration = require('./configuration/configuration')
+const extractValidationStats = require('./stats/extractValidationStats')
+
+const SUPPORTED_VALIDATOR = [
+  'title',
+  'label',
+  'description',
+  'assignee',
+  'dependent',
+  'milestone',
+  'approval'
+]
+
+const SUPPORTED_ACTIONS = [
+  'comments',
+  'checks'
+]
+
+// Main logic Processor of mergeable
+const executeMergeable = async (context, registry) => {
+  if (registry === undefined) {
+    registry = { validators: new Map(), actions: new Map() }
+  }
+
+  // first fetch the configuration
+  let config = await Configuration.instanceWithContext(context)
+
+  // check config file and get settings only for the current event. the rest is useless for executor
+  let eventName = `${context.event}.${context.payload.action}`
+  config = filterByEvent(config, eventName)
+  // go through the settings and register all the validators
+  registerValidatorsAndActions(config, registry)
+
+  // do pre validation actions
+  await processPreActions(context, registry, config)
+  config.forEach(rule => {
+    Promise.all(getValiationPromises(context, registry, rule))
+      .then((result) => {
+        const translatedOutput = extractValidationStats(result)
+
+        Promise.all(getActionPromises(context, registry, rule, translatedOutput))
+      })
+  })
+}
+
+const filterByEvent = (config, eventName) => {
+  let filteredConfig = []
+  config.settings.mergeable.forEach(rule => {
+    let events = Array.isArray(rule.when) ? rule.when : rule.when.split(',').map(n => n.trim())
+    if (isEventInContext(eventName, events)) {
+      filteredConfig.push(rule)
+    }
+  })
+
+  return filteredConfig
+}
+
+const isEventInContext = (eventName, events) => {
+  let eventObject = eventName.split('.')[0]
+  return ((events.indexOf(eventName.trim()) > -1) ||
+    (events.indexOf('pull_request.*') > -1 && eventObject === 'pull_request') ||
+    (events.indexOf('issues.*') > -1 && eventObject === 'issues'))
+}
+
+const registerValidatorsAndActions = (config, registry) => {
+  config.forEach(rule => {
+    rule.validate.forEach(validation => {
+      let key = validation.do
+      if (!registry.validators.has(key)) {
+        if (!SUPPORTED_VALIDATOR.includes(key)) {
+          return // add validator not supported error
+        }
+
+        let Validator = require(`./validators/${key}`)
+        registry.validators.set(key, new Validator())
+      }
+    })
+    rule.fail.concat(rule.pass).forEach(action => {
+      let key = action.do
+      if (!registry.actions.has(key)) {
+        if (!SUPPORTED_ACTIONS.includes(key)) {
+          return // add validator not supported error
+        }
+
+        let Action = require(`./actions/${key}`)
+        registry.actions.set(key, new Action())
+      }
+    })
+  })
+}
+
+const getValiationPromises = (context, registry, rule) => {
```

Comment:
Should be called `getValidatorPormises` IMO since our classes are named `Validators`

Reply:
done